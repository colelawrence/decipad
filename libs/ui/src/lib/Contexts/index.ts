export * from './ProgramBlocks';
export * from './ExternalAuthentication';
