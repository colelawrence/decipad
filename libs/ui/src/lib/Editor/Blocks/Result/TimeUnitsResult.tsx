import plural from 'pluralize';

export function TimeUnitsResult({
  value: entries,
}: {
  value: [string, number][];
}): JSX.Element {
  return (
    <>
      [ {entries.map(([name, val]) => `${val} ${plural(name, val)}`).join(', ')}{' '}
      ]
    </>
  );
}
