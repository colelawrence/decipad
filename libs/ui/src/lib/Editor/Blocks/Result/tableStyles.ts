import { chakra } from '@chakra-ui/system';
import { Box, Table, Tr, Th, Td } from '@chakra-ui/react';

export const TableHeader = chakra(Th, {
  baseStyle: {
    textTransform: 'none',
    fontSize: 'md',
    borderTop: 0,
    borderRight: '1px',
    borderColor: 'gray.100',
    color: 'black',
    wordWrap: 'no-wrap',
    bg: 'transparent',
    _last: {
      borderRight: 0,
    },
  },
});

export const HeaderIcon = chakra(Box, {
  baseStyle: {
    d: 'inline',
    color: 'gray.500',
    marginRight: '0.5em',
  },
});

export const TableRow = chakra(Tr, {
  baseStyle: {
    borderBottom: '1px',
    borderBottomColor: 'gray.100',
    _last: {
      borderBottom: 0,
    },
  },
});

export const TableCell = chakra(Td, {
  baseStyle: {
    bg: 'white',
    borderRight: '1px',
    borderColor: 'gray.100',
    px: 3,
    py: 2.5,
    borderBottom: 0,
    _last: {
      borderRight: 0,
    },
  },
});

export const ResultTable = chakra(Table, {
  baseStyle: {
    tableLayout: 'fixed',
  },
});
