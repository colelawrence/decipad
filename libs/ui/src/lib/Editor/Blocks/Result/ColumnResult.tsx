import { Box, Tbody } from '@chakra-ui/react';
import { Interpreter } from '@decipad/language';
import { InlineColumnResult } from '../../../../organisms';
import { ResultTable, TableRow, TableCell } from './tableStyles';
import { ResultContent } from './Result.component';
import { ResultTypeProps } from '../../../results';
import { useResults } from '../../../Contexts/Results';

export function ColumnResult({
  type,
  value,
  variant = 'block',
}: ResultTypeProps): JSX.Element | null {
  const { indexLabels } = useResults();

  if (variant === 'inline') {
    return <InlineColumnResult type={type} value={value} />;
  }

  const { indexedBy, cellType } = type;

  if (!cellType) return null;

  const border = {
    borderTopRadius: 6,
    border: '1px',
    borderColor: 'gray.100',
  };

  return (
    <Box {...border}>
      <ResultTable role="table">
        {/* TODO: Column caption should say the name of the variable (if there is one. */}
        <Tbody>
          {(value as Interpreter.ResultColumn).map((row, rowIndex) => {
            const labels = indexLabels.get(indexedBy ?? '');

            return (
              <TableRow key={rowIndex}>
                {labels && (
                  <TableCell isNumeric={false}>
                    <span style={{ color: 'gray' }}>{labels[rowIndex]}</span>
                  </TableCell>
                )}
                <TableCell isNumeric={cellType.type === 'number'}>
                  <ResultContent type={cellType} value={row} />
                </TableCell>
              </TableRow>
            );
          })}
        </Tbody>
      </ResultTable>
    </Box>
  );
}
