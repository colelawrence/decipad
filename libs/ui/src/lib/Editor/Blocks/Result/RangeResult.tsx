import { ResultContent, ResultContentProps } from './Result.component';

export function RangeResult({
  type: { rangeOf },
  value,
}: ResultContentProps): JSX.Element | null {
  if (!rangeOf) {
    return null;
  }

  return (
    <>
      <ResultContent type={rangeOf} value={value[0]} /> &rarr;{' '}
      <ResultContent type={rangeOf} value={value[1]} />
    </>
  );
}
