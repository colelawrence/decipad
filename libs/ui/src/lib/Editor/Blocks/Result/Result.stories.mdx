import { Meta, Story, Canvas, Props } from '@storybook/addon-docs/blocks';

import { Type } from '@decipad/language';
import { Result } from './Result.component';

<Meta
  title="Result"
  parameters={{
    docs: {
      source: {
        code: `<Result type={/* Type class from @decipad/language */} value={[[1234]]} />`,
      },
    },
  }}
  component={Result}
/>

# Result

The Result component is one of the most important components for our editor, this is where the user will do all of their calculations and decisions. The Result component gets automatic syntax highlighting once it is plugged into slate. The component itself is a slate specific component where it requires a result to be either undefined or a string to output the calculation's result.

export const Template = (args) => <Result {...args} />;

<Canvas>
  <Story
    name="Numeric result"
    args={{
      value: [1234],
      type: Type.build({ type: 'number' }),
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="String result"
    args={{
      value: ['ayyy'],
      type: Type.build({ type: 'string' }),
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="Column result"
    args={{
      value: [[1, 2, 3]],
      type: Type.buildColumn(Type.build({ type: 'number' }), 3),
    }}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="Errored result"
    args={{
      value: [[1, 2, 3]],
      type: Type.Impossible.withErrorCause('Oops! Some error occurred'),
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>
