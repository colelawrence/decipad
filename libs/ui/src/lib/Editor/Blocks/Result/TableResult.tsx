import { Box, Tbody, Thead } from '@chakra-ui/react';
import { Type } from '@decipad/language';
import { FiCalendar, FiHash, FiType, FiHelpCircle } from 'react-icons/fi';

import {
  ResultTable,
  TableRow,
  TableCell,
  TableHeader,
  HeaderIcon,
} from './tableStyles';
import { ResultContent, ResultContentProps } from './Result.component';

function IconForType({ type }: { type: Type }): JSX.Element {
  let Icon;
  if (type.date) {
    Icon = FiCalendar;
  } else if (type.type === 'number') {
    Icon = FiHash;
  } else if (type.type === 'boolean') {
    Icon = FiHelpCircle;
  } else {
    Icon = FiType;
  }

  return (
    <HeaderIcon>
      <Icon style={{ display: 'inline', verticalAlign: 'text-top' }} />
    </HeaderIcon>
  );
}

function tableByRows(
  columnNames: string[],
  columnTypes: Type[],
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  columnsOrCells: any[]
) {
  const rows = [];
  for (let index = 0; index < columnsOrCells[0]?.length ?? 0; index += 1) {
    const row = [];
    for (const column of columnsOrCells) {
      row.push(column[index]);
    }
    rows.push(row);
  }

  return { columnTypes, columnNames, rows };
}

export function TableResult({
  type: { columnNames, columnTypes },
  value,
  depth = 0,
}: ResultContentProps): JSX.Element | null {
  if (!columnNames || !columnTypes) return null;

  const table = tableByRows(columnNames, columnTypes, value);
  const border = {
    borderTopRadius: 6,
    border: '1px',
    borderColor: 'gray.100',
  };

  return (
    <Box {...border}>
      <ResultTable>
        {/* TODO: table caption should say the name of the variable (if there is one. */}
        <Thead>
          <TableRow>
            {table.columnNames.map((columnName, colIndex) => {
              const t = table.columnTypes[colIndex];

              return (
                <TableHeader key={colIndex}>
                  <IconForType type={t} />
                  &nbsp;{columnName}
                </TableHeader>
              );
            })}
          </TableRow>
        </Thead>
        <Tbody>
          {table.rows.map((row, rowIndex) => {
            return (
              <TableRow key={rowIndex}>
                {row.map((cell, index) => {
                  const t = table.columnTypes[index];
                  return (
                    <TableCell key={index} isNumeric={t.type === 'number'}>
                      <ResultContent
                        key={index}
                        type={t}
                        value={cell}
                        depth={depth + 1}
                      />
                    </TableCell>
                  );
                })}
              </TableRow>
            );
          })}
        </Tbody>
      </ResultTable>
    </Box>
  );
}
