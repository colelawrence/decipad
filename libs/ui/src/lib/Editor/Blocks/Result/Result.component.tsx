/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { chakra } from '@chakra-ui/system';
import { Box } from '@chakra-ui/react';
import { Type, InBlockResult, OptionalValueLocation } from '@decipad/language';
import { useResults } from '@decipad/ui';

import { DateResult, NumberResult, TimeUnitsResult } from '../../../../atoms';
import { RangeResult } from '../../../../organisms';

import { ColumnResult } from './ColumnResult';
import { TableResult } from './TableResult';

const commonStyles = {
  py: 2,
  px: 6,
  d: 'inline-block',
  borderBottomRadius: 6,
  mb: 3,
};

export const ResultErrorStyles = chakra(Box, {
  baseStyle: {
    bg: 'red.100',
    color: 'red.500',
    fontWeight: 'bold',
    ...commonStyles,
  },
});

export const DefaultResultStyles = chakra(Box, {
  baseStyle: {
    bg: 'rgb(252,252,263)',
    color: 'black',
    ...commonStyles,
    w: '100%',
  },
});

export interface ResultContentProps {
  type: Type;
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  value: any | any[];
  depth?: number;
}

export const ResultContent = (props: ResultContentProps) => {
  const { type, value } = props;
  if (type == null) return null;

  if (type.type === 'number') {
    return <NumberResult {...props} />;
  }
  if (type.type === 'boolean' || type.type === 'string') {
    return <>{String(value)}</>;
  }
  if (type.date) {
    return <DateResult {...props} />;
  }
  if (type.columnTypes != null) {
    return <TableResult {...props} />;
  }
  if (type.columnSize != null && Array.isArray(value)) {
    return <ColumnResult {...props} />;
  }
  if (type.functionness) {
    return <>Æ’</>;
  }
  if (type.timeUnits) {
    return <TimeUnitsResult {...props} />;
  }
  if (type.rangeOf) {
    return <RangeResult {...props} />;
  }
  return value.toString();
};

const getLineResult = (
  currentBlockId: string,
  cursor: OptionalValueLocation | null,
  results: InBlockResult[]
) => {
  let underCursor = null;

  if (cursor != null) {
    underCursor = results.find(
      ({ blockId, statementIndex }) =>
        blockId === cursor[0] && statementIndex === cursor[1]
    );
  }

  if (currentBlockId === cursor?.[0] && cursor?.[1] === null) {
    return null;
  }

  return underCursor ?? results[results.length - 1];
};

const NilResult = () => (
  <DefaultResultStyles contentEditable={false}>&nbsp;</DefaultResultStyles>
);

export const Result = ({
  blockId,
  useDefaultStyles = true,
}: {
  blockId: string;
  useDefaultStyles?: boolean;
}): JSX.Element | null => {
  const { cursor, blockResults } = useResults();

  const blockResult = blockResults[blockId];

  if (blockResult == null) {
    return <NilResult />;
  }

  if (blockResult.isSyntaxError) {
    return <ResultErrorStyles contentEditable={false}>Error</ResultErrorStyles>;
  }

  if (blockResult.results.length === 0) {
    return <NilResult />;
  }

  const lineResult = getLineResult(blockId, cursor, blockResult.results);
  if (lineResult?.valueType.errorCause != null) {
    return (
      <ResultErrorStyles contentEditable={false}>
        {lineResult?.valueType.errorCause.message}
      </ResultErrorStyles>
    );
  }
  if (lineResult?.value != null) {
    const result = (
      <ResultContent type={lineResult.valueType} value={lineResult.value} />
    );
    const resultWithStyles = useDefaultStyles ? (
      <DefaultResultStyles>{result}</DefaultResultStyles>
    ) : (
      result
    );
    return <div contentEditable={false}>{resultWithStyles}</div>;
  }
  return <NilResult />;
};
