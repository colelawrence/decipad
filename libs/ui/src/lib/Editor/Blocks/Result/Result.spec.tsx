import { render } from 'test-utils';
import { Type } from '@decipad/language';
import { Result } from './Result.component';
import { IdentifiedResult } from 'libs/language/src/computer/types';
import { ResultsContextProvider, ResultsContextValue } from '../../../Contexts';

const renderResult = (result: IdentifiedResult) => {
  const allResults: IdentifiedResult[] = [
    {
      blockId: 'wrongId',
      isSyntaxError: true,
      results: [],
    },
    result,
  ];

  const contextValue: ResultsContextValue = {
    cursor: null,
    blockResults: Object.fromEntries(allResults.map((b) => [b.blockId, b])),
  };

  return render(
    <ResultsContextProvider value={contextValue}>
      <Result blockId="blockUnderTest" />
    </ResultsContextProvider>
  );
};

describe('Result Block', () => {
  it('renders', () => {
    const { container } = renderResult({
      blockId: 'blockUnderTest',
      isSyntaxError: false,
      results: [
        {
          blockId: 'blockUnderTest',
          statementIndex: 0,
          valueType: Type.Number,
          value: 3,
        },
      ],
    });

    expect(container).toBeDefined();
  });

  it('matches snapshot for a number', () => {
    const { container } = renderResult({
      blockId: 'blockUnderTest',
      isSyntaxError: false,
      results: [
        {
          blockId: 'blockUnderTest',
          statementIndex: 0,
          valueType: Type.Number,
          value: 3,
        },
      ],
    });

    expect(container).toMatchSnapshot();
  });

  it('matches snapshot for a column', () => {
    const { container } = renderResult({
      blockId: 'blockUnderTest',
      isSyntaxError: false,
      results: [
        {
          blockId: 'blockUnderTest',
          statementIndex: 0,
          valueType: Type.buildColumn(Type.Number, 3),
          value: [1, 2, 3],
        },
      ],
    });

    expect(container).toMatchSnapshot();
  });

  it('matches snapshot for a column of columns', () => {
    const { container } = renderResult({
      blockId: 'blockUnderTest',
      isSyntaxError: false,
      results: [
        {
          blockId: 'blockUnderTest',
          statementIndex: 0,
          valueType: Type.buildColumn(Type.buildColumn(Type.String, 2), 3),
          value: [
            ['1', '2'],
            ['3', '4'],
            ['5', '6'],
          ],
        },
      ],
    });

    expect(container).toMatchSnapshot();
  });

  it('matches snapshot for a syntax error', () => {
    const { container } = renderResult({
      blockId: 'blockUnderTest',
      isSyntaxError: true,
      results: [],
    });

    expect(container).toMatchSnapshot();
  });
});
