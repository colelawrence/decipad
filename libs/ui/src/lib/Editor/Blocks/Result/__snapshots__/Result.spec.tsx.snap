// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result Block matches snapshot for a column 1`] = `
.emotion-0 {
  background: rgb(252,252,263);
  color: var(--chakra-colors-black);
  padding-top: var(--chakra-space-2);
  padding-bottom: var(--chakra-space-2);
  -webkit-padding-start: var(--chakra-space-6);
  padding-inline-start: var(--chakra-space-6);
  -webkit-padding-end: var(--chakra-space-6);
  padding-inline-end: var(--chakra-space-6);
  display: inline-block;
  border-bottom-left-radius: 6px;
  border-bottom-right-radius: 6px;
  margin-bottom: var(--chakra-space-3);
  width: 100%;
}

<div>
  <div
    class="emotion-0"
    contenteditable="false"
  >
    [ 
    1
    
    , 
    2
    
    , 
    3
    
     ]
  </div>
</div>
`;

exports[`Result Block matches snapshot for a column of columns 1`] = `
.emotion-0 {
  background: rgb(252,252,263);
  color: var(--chakra-colors-black);
  padding-top: var(--chakra-space-2);
  padding-bottom: var(--chakra-space-2);
  -webkit-padding-start: var(--chakra-space-6);
  padding-inline-start: var(--chakra-space-6);
  -webkit-padding-end: var(--chakra-space-6);
  padding-inline-end: var(--chakra-space-6);
  display: inline-block;
  border-bottom-left-radius: 6px;
  border-bottom-right-radius: 6px;
  margin-bottom: var(--chakra-space-3);
  width: 100%;
}

<div>
  <div
    class="emotion-0"
    contenteditable="false"
  >
    [ 
    [ 
    1
    , 
    2
     ]
    , 
    [ 
    3
    , 
    4
     ]
    , 
    [ 
    5
    , 
    6
     ]
     ]
  </div>
</div>
`;

exports[`Result Block matches snapshot for a number 1`] = `
.emotion-0 {
  background: rgb(252,252,263);
  color: var(--chakra-colors-black);
  padding-top: var(--chakra-space-2);
  padding-bottom: var(--chakra-space-2);
  -webkit-padding-start: var(--chakra-space-6);
  padding-inline-start: var(--chakra-space-6);
  -webkit-padding-end: var(--chakra-space-6);
  padding-inline-end: var(--chakra-space-6);
  display: inline-block;
  border-bottom-left-radius: 6px;
  border-bottom-right-radius: 6px;
  margin-bottom: var(--chakra-space-3);
  width: 100%;
}

<div>
  <div
    class="emotion-0"
    contenteditable="false"
  >
    3
    
  </div>
</div>
`;
